cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java

```diff
@@ -15,44 +15,260 @@
+import co.cask.cdap.data2.transaction.queue.QueueConstants;
 import co.cask.cdap.data2.transaction.queue.QueueEntryRow;
-import org.apache.hadoop.hbase.client.Get;
-import org.apache.hadoop.hbase.client.HTable;
-import org.apache.hadoop.hbase.client.Put;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Longs;
 
-import java.io.IOException;
+import java.util.Map;
 
 /**
  * Class for persisting consumer state per queue consumer.
  */
-final class HBaseConsumerStateStore {
+final class HBaseConsumerStateStore extends AbstractDataset implements QueueConfigurer {
 
   private final QueueName queueName;
-  private final ConsumerConfig consumerConfig;
-  private final HTable hTable;
+  private final Table table;
+  private final byte[] minStartRow;
 
-  HBaseConsumerStateStore(QueueName queueName, ConsumerConfig consumerConfig, HTable hTable) {
+  HBaseConsumerStateStore(QueueName queueName, Table table) {
+    super(Constants.SYSTEM_NAMESPACE + "." + QueueConstants.QueueType.QUEUE, table);
     this.queueName = queueName;
-    this.consumerConfig = consumerConfig;
-    this.hTable = hTable;
+    this.table = table;
+    this.minStartRow = QueueEntryRow.getQueueEntryRowKey(queueName, 0L, 0);
   }
 
   /**
-   * Returns the start row as stored in the state store.
+   * Returns the consumer state as stored in the state store for the given consumer.
    */
-  public HBaseConsumerState getState() throws IOException {
-    Get get = new Get(queueName.toBytes());
-    byte[] column = HBaseQueueAdmin.getConsumerStateColumn(consumerConfig.getGroupId(), consumerConfig.getInstanceId());
-    get.addColumn(QueueEntryRow.COLUMN_FAMILY, column);
+  HBaseConsumerState getState(long groupId, int instanceId) {
+    byte[] stateColumn = getConsumerStateColumn(groupId, instanceId);
+    byte[] startRow = table.get(queueName.toBytes(), stateColumn);
+    Preconditions.checkState(startRow != null,
+                             "No consumer state found. GroupId: %s, InstanceId: %s", groupId, instanceId);
 
-    return new HBaseConsumerState(hTable.get(get), consumerConfig.getGroupId(), consumerConfig.getInstanceId());
+    return new HBaseConsumerState(startRow, groupId, instanceId);
   }
 
-  public void saveState(HBaseConsumerState state) throws IOException {
-    // Writes latest startRow to queue config.
-    hTable.put(state.updatePut(new Put(queueName.toBytes())));
-    hTable.flushCommits();
+  /**
+   * Updates the start row state of the given consumer.
+   */
+  void updateState(long groupId, int instanceId, byte[] startRow) {
+    byte[] stateColumn = getConsumerStateColumn(groupId, instanceId);
+    table.put(queueName.toBytes(), stateColumn, startRow);
```

Comment:
`QueueName` is doing that already, hence the `toByte()` call is just retuning a `byte[]`.